test/parallel/test-fs-read-position-validation.mjs

```diff
@@ -0,0 +1,99 @@
+import * as common from '../common/index.mjs';
+import * as fixtures from '../common/fixtures.mjs';
+import fs from 'fs';
+import assert from 'assert';
+
+// This test ensures that "position" argument is correctly validated
+
+const filepath = fixtures.path('x.txt');
+
+const buffer = Buffer.from('xyz\n');
+const offset = 0;
+const length = buffer.byteLength;
+
+async function testValid(position) {
+  let fdSync;
+  fs.open(filepath, 'r', common.mustSucceed((fd) => {
+    try {
+      fs.read(fd, buffer, offset, length, position, common.mustSucceed());
+      fs.read(fd, { buffer, offset, length, position }, common.mustSucceed());
+    } finally {
+      fs.close(fd, common.mustSucceed());
+    }
+  }));
+  try {
+    fdSync = fs.openSync(filepath, 'r');
+    fs.readSync(fdSync, buffer, offset, length, position);
+    fs.readSync(fdSync, buffer, { offset, length, position });
+  } finally {
+    if (fdSync) fs.closeSync(fdSync);
+  }
+}
+
+async function testInvalid(code, position, skipCb = false) {
+  let fdSync;
+  try {
+    fdSync = fs.openSync(filepath, 'r');
+    assert.throws(
+      () => fs.readSync(fdSync, buffer, offset, length, position),
+      { code }
+    );
+    assert.throws(
+      () => fs.readSync(fdSync, buffer, { offset, length, position }),
+      { code }
+    );
+  } finally {
+    if (fdSync) fs.closeSync(fdSync);
+  }
+
+  // Allows to skip Callback API test for uncatchable cases (for example, EINVAL)
+  if (skipCb) return;
+
+  fs.open(filepath, 'r', common.mustSucceed((fd) => {
+    try {
+      assert.throws(
+        () => fs.read(fd, buffer, offset, length, position, common.mustNotCall()),
+        { code }
+      );
+      assert.throws(
+        () => fs.read(fd, { buffer, offset, length, position }, common.mustNotCall()),
+        { code }
+      );
+    } finally {
+      fs.close(fd, common.mustSucceed());
+    }
+  }));
+}
+
+{
+  await testValid(undefined);
+  await testValid(null);
+  await testValid(-1);
+  await testValid(-1n);
+
+  await testValid(0);
+  await testValid(0n);
+  await testValid(1);
+  await testValid(1n);
+  await testValid(9);
+  await testValid(9n);
+  await testValid(Number.MAX_SAFE_INTEGER);
+
+  await testValid(2n ** 63n - 1n - BigInt(length));
+  await testInvalid('ERR_OUT_OF_RANGE', 2n ** 63n);
+  await testInvalid('ERR_OUT_OF_RANGE', 2n ** 63n - BigInt(length));
```

Comment:
It looks like passing `2n ** 63n - BigInt(length)` works on `master` (on both the sync and the async version), I'm not seeing any EINVAL error 🤔 maybe it's OS specific or something? If that's the case, it's probably not worth doing it in a separate PR, it's probably easier to land it alongside the other changes as semver-major.

Reply:
Huh. Well, it definitely depends on `read` syscall implementation.
I retried this on three Node.js versions:
```mjs
import { fileURLToPath } from 'url';
import { openSync, readSync } from 'fs';
const fd = openSync(fileURLToPath(import.meta.url));
const buffer = Buffer.alloc(9);
readSync(fd, buffer, { position: 2n ** 63n - 15n, length: 9 }); console.log(buffer); // should work
readSync(fd, buffer, { position: 2n ** 63n - 5n, length: 9 }); console.log(buffer); // might fail with EINVAL
readSync(fd, buffer, { position: 2n ** 63n + 5n, length: 9 }); console.log(buffer); // must fail with exception
```
v14.17.6 most likely doesn't support bigints and assumes current position so it reads from the start:
```
<Buffer 69 6d 70 6f 72 74 20 7b 20>
<Buffer 66 69 6c 65 55 52 4c 54 6f>
<Buffer 50 61 74 68 20 7d 20 66 72>
```
v16.14.1 and master (v19.0.0-pre) both show this:
```
<Buffer 00 00 00 00 00 00 00 00 00>
node:fs:732
  handleErrorFromBinding(ctx);
  ^

Error: EINVAL: invalid argument, read
    at readSync (node:fs:732:3)
    at file:///tmp/chk.mjs:6:1
    at ModuleJob.run (node:internal/modules/esm/module_job:198:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:409:24)
    at async loadESM (node:internal/process/esm_loader:85:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12) {
  errno: -22,
  syscall: 'read',
  code: 'EINVAL'
}
```
This is on 64bit linux 5.17.0.

If the `- 5n` line works on any other platform, it is definitely semver-major indeed.

Reply:
Tried v18.0.0 release build on 64bit windows 10.0.18363, same result
```cmd
>node.exe c:\chk.mjs
<Buffer 00 00 00 00 00 00 00 00 00>
node:fs:732
  handleErrorFromBinding(ctx);
  ^

Error: EINVAL: invalid argument, read
    at readSync (node:fs:732:3)
    at file:///c:/chk.mjs:6:1
    at ModuleJob.run (node:internal/modules/esm/module_job:198:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:409:24)
    at async loadESM (node:internal/process/esm_loader:85:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12) {
  errno: -4071,
  syscall: 'read',
  code: 'EINVAL'
}

Node.js v18.0.0
```
Out of curiosity, on which platform it works successfully?

Reply:
I’ve tested it on macOS. Could we try to revert the changes in lib, change the test case to catch EINVAL, and run the CI to see which platforms behave like this? (in case that’s just an oddity of my machine or if I’ve tested it wrong somehow)

Reply:
So it looks like the behavior is indeed platform specific. I suggest we keep in this PR all the tests that pass on `master` and make the behavior consistent in the other PR that's semver-major.

Reply:
Yep! Can we have one more CI run please? I wanna make sure that testing this part for success and continuing on won't show anything unexpected for these platforms.

Also, results on AIX are bothering: https://ci.nodejs.org/job/node-test-commit-aix/40804/nodes=aix72-ppc64/testReport/junit/(root)/test/parallel_test_fs_read_position_validation/
This one reported `Error: EFBIG: file too large, read` on the test with `Number.MAX_SAFE_INTEGER`
I guess there's no choice but to suppress this particular error in testValid.

Reply:
Thanks a lot.
If I'm reading CI reports correctly, the results were mutually exclusive, except for AIX.
Helper functions got slightly overgeneralized but it should be fine.